### PR TITLE
Catch SameFileError errors when contructing RPath on Windows

### PR DESCRIPTION
**Pull request by johnwparent**
Improve error handling during link phase of Windows simulated RPaths. The symlink module, with symlinks disabled on Windows, will defer to junctions. Junctions first test if the link is viable by checking if the error already exists. For some reason in this scenario, we were previously trying to let  `shutil` copy the file, which we know will fail. This PR improves the error handling by catching the SameFileError from `shutil` and additionally, skips that code path altogether if we know we'll hit an `EEXIST` error.

In short this PR avoids an unnecessary operation, and improves the RPath error handling.

### CODE REVIEW

**Comment by johnwparent** [comment]
@kwryankrattiger this should resolve your Perl issue.

File: lib/spack/llnl/util/filesystem.py [hunk 1]
```diff
@@ -2407,7 +2407,11 @@ def _link(self, path, dest_dir):
             # For py2 compatibility, we have to catch the specific Windows error code
             # associate with trying to create a file that already exists (winerror 183)
             except OSError as e:
-                if sys.platform == "win32" and e.winerror == 183:
+                if sys.platform == "win32" and (
+                    e.winerror == 183
+                    or e.errno == errno.EEXIST
```
Comment (scheibelp) [inline lib/spack/llnl/util/filesystem.py]: For this check, isn't code 183 for when the file already exists? Are there other pathways on Windows that were resulting in this `errno`? AFAIK `errno` is Linux-specific so I'm surprised we'd be seeing it at all.

File: lib/spack/llnl/util/filesystem.py [hunk 2]
```diff
@@ -2407,7 +2407,11 @@ def _link(self, path, dest_dir):
...
+                    or isinstance(e, shutil.SameFileError)
```
Comment (scheibelp) [inline lib/spack/llnl/util/filesystem.py]: It's odd that we'd be seeing this given that this is attempted inside of `if os.path.exists(dest_dir) and not dest_file == path:`. Is that on account of a lack of resolution of symlinks etc. to a normalized, absolute path?

File: lib/spack/llnl/util/symlink.py [hunk 3]
```diff
@@ -30,9 +30,15 @@ def symlink(real_path, link_path):
...
             _win32_junction(real_path, link_path)
-        except OSError:
-            # If all else fails, fall back to copying files
-            shutil.copyfile(real_path, link_path)
+        except OSError as e:
+            if e.errno == errno.EEXIST:
+                # we manually raised this error, we know shutil will fail the copy
+                # due to a same file error
+                # skip and report
+                raise e
```
Comment (scheibelp) [inline lib/spack/llnl/util/symlink.py]: Use `raise` rather than `raise e` (the latter interferes with the traceback)

File: lib/spack/llnl/util/symlink.py [hunk 4]
```diff
@@ -30,9 +30,15 @@ def symlink(real_path, link_path):
...
             # Try to use junctions
             _win32_junction(real_path, link_path)
-        except OSError:
-            # If all else fails, fall back to copying files
-            shutil.copyfile(real_path, link_path)
+        except OSError as e:
+            if e.errno == errno.EEXIST:
+                # we manually raised this error, we know shutil will fail the copy
+                # due to a same file error
+                # skip and report
```
Comment (scheibelp) [inline lib/spack/llnl/util/symlink.py]: IMO it's more important that the error explains that we don't need to perform this operation, vs. explaining that we know it was raised manually (I would avoid that kind of explanation in general, including use of the term "manual")

**Review by scheibelp — CHANGES_REQUESTED** [review]
I have a couple questions and a couple suggestions.

File: lib/spack/llnl/util/filesystem.py (hunk 1 above)
Comment (johnwparent) [reply lib/spack/llnl/util/filesystem.py]: We set `e.errno` explicitly in `llnl/util/symlink.py::_win32_juction` in this scenario, but in general you are correct.

File: lib/spack/llnl/util/symlink.py (hunk 3 above)
Comment (johnwparent) [reply lib/spack/llnl/util/symlink.py]: Huh, I did not know that... TMYK. Amended.

File: lib/spack/llnl/util/symlink.py (hunk 4 above)
Comment (johnwparent) [reply lib/spack/llnl/util/symlink.py]: I'll refactor the explanation along those lines.

File: lib/spack/llnl/util/filesystem.py (hunk 2 above)
Comment (johnwparent) [reply lib/spack/llnl/util/filesystem.py]: `not dest_file == path` will only be true if we're symlinking something like `<spack_root>/opt/path-to-package/bin/zlib.dll` to itself, as all paths are absolute. This does not help us in a scenario where we're linking something from somewhere like `<spack_root>/opt/path-to-package/x64/bin/zlib.dll` to `<spack_root>/opt/path-to-package/bin/zlib.dll`, where there will be a collision if `zlib.dll` is already in the bin as those paths are not equivalent, and that if statement will allow this through.

Comment (johnwparent) [reply lib/spack/llnl/util/filesystem.py]: We could check the destination ourselves in cases like this and avoid the error handling altogether.

Comment (scheibelp) [reply lib/spack/llnl/util/filesystem.py]: According to https://docs.python.org/3/library/shutil.html, `SameFileError` is only supposed to be raised if "source and destination in [copyfile()](https://docs.python.org/3/library/shutil.html#shutil.copyfile) are the same file."

If `<spack_root>/opt/path-to-package/bin/zlib.dll` already exists, I would not expect `SameFileError` to be raised, but rather an `IOError` with one of the already-checked error codes (e.g. `EEXIST`). Are you saying that `shutil` is raising this error in that case though?

Comment (johnwparent) [reply lib/spack/llnl/util/filesystem.py]: Yes, attached is the screenshot I've been working from to diagnose this: 
unnamed

Comment (johnwparent) [reply lib/spack/llnl/util/filesystem.py]: I think from `shutil`'s perspective, they are the "same file"

**Comment by johnwparent** [comment]
@spackbot fix style

**Review by scheibelp — APPROVED** [review]
